sonar-batch/src/main/java/org/sonar/batch/report/TestExecutionAndCoveragePublisher.java

```diff
@@ -0,0 +1,128 @@
+
+  private final ResourceCache resourceCache;
+  private final TestPlanBuilder testPlanBuilder;
+
+  public TestExecutionAndCoveragePublisher(ResourceCache resourceCache, TestPlanBuilder testPlanBuilder) {
+    this.resourceCache = resourceCache;
+    this.testPlanBuilder = testPlanBuilder;
+  }
+
+  @Override
+  public void publish(BatchReportWriter writer) {
+    for (final BatchResource resource : resourceCache.all()) {
+      if (!resource.isFile()) {
+        continue;
+      }
+
+      DefaultInputFile inputFile = (DefaultInputFile) resource.inputPath();
+      if (inputFile.type() != Type.TEST) {
+        continue;
+      }
+
+      final MutableTestPlan testPlan = testPlanBuilder.get(MutableTestPlan.class, inputFile.key());
+      if (testPlan == null || Iterables.isEmpty(testPlan.testCases())) {
+        continue;
+      }
+
+      final Set<String> testNamesWithCoverage = new HashSet<>();
+
+      writer.writeTests(resource.batchId(), Iterables.transform(testPlan.testCases(), new Function<MutableTestCase, BatchReport.Test>() {
+
+        private BatchReport.Test.Builder builder = BatchReport.Test.newBuilder();
+
+        @Override
+        public Test apply(MutableTestCase testCase) {
+          builder.clear();
+          builder.setName(testCase.name());
+          if (testCase.doesCover()) {
+            testNamesWithCoverage.add(testCase.name());
+          }
+          Long durationInMs = testCase.durationInMs();
+          if (durationInMs != null) {
+            builder.setDurationInMs(durationInMs);
+          }
+          String msg = testCase.message();
+          if (msg != null) {
+            builder.setMsg(msg);
+          }
+          String stack = testCase.stackTrace();
+          if (stack != null) {
+            builder.setStacktrace(stack);
+          }
+          TestCase.Status status = testCase.status();
+          if (status != null) {
+            builder.setStatus(TestStatus.valueOf(status.name()));
+          }
+          return builder.build();
+        }
+
+      }));
+
+      writer.writeCoverageDetails(resource.batchId(), Iterables.transform(testNamesWithCoverage, new Function<String, CoverageDetail>() {
+
+        private BatchReport.CoverageDetail.Builder builder = BatchReport.CoverageDetail.newBuilder();
+        private BatchReport.CoverageDetail.CoveredFile.Builder coveredBuilder = BatchReport.CoverageDetail.CoveredFile.newBuilder();
+
+        @Override
+        public CoverageDetail apply(String testName) {
+          // Take first test with provided name
+          MutableTestCase testCase = testPlan.testCasesByName(testName).iterator().next();
+          builder.clear();
+          builder.setTestName(testName);
+          for (CoverageBlock block : testCase.coverageBlocks()) {
+            coveredBuilder.clear();
+            coveredBuilder.setFileRef(resourceCache.get(block.testable().component().key()).batchId());
```

Comment:
:new: ![CRITICAL](https://raw.githubusercontent.com/henryju/image-hosting/master/critical.png)Possible null pointer dereference in org.sonar.batch.report.TestExecutionAndCoveragePublisher$2.apply(String) due to return value of called method[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)